File: backend/app/utils/transaction_coordinator.py
"""
Backend Transaction Coordinator
Handles database transactions with proper ordering to prevent IRR calculation race conditions
"""

import logging
from typing import List, Dict, Any, Optional, Tuple
from datetime import datetime
from app.db.database import get_db
from app.api.routes.portfolio_funds import calculate_single_portfolio_fund_irr

logger = logging.getLogger(__name__)

class TransactionCoordinator:
    """
    Coordinates database transactions to ensure proper ordering of activities and valuations
    """
    
    def __init__(self, db):
        self.db = db
        
    async def save_activities_and_valuations_ordered(
        self,
        activities: List[Dict[str, Any]],
        valuations: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """
        Save activities and valuations in proper order with transaction safety
        
        Args:
            activities: List of activity data to save
            valuations: List of valuation data to save
            
        Returns:
            Dictionary with transaction results
        """
        
        result = {
            "success": True,
            "activities_saved": 0,
            "valuations_saved": 0,
            "irr_calculations": 0,
            "portfolio_irr_recalculations": 0,
            "errors": []
        }
        
        try:
            logger.info(f"🔄 Backend Transaction: Starting ordered save of {len(activities)} activities and {len(valuations)} valuations")
            
            # Phase 1: Save all activities first
            if activities:
                logger.info("📥 Phase 1: Saving activities...")
                for activity in activities:
                    await self._save_activity(activity)
                    result["activities_saved"] += 1
                logger.info(f"✅ Phase 1 Complete: {result['activities_saved']} activities saved")
            
            # Phase 2: Save valuations after activities
            if valuations:
                logger.info("💰 Phase 2: Saving valuations...")
                for valuation in valuations:
                    await self._save_valuation(valuation)
                    result["valuations_saved"] += 1
                logger.info(f"✅ Phase 2 Complete: {result['valuations_saved']} valuations saved")
            
            # Phase 3: Trigger IRR recalculation for affected funds (both fund-level and portfolio-level)
            affected_funds = self._get_affected_funds(activities, valuations)
            affected_portfolios = set()  # Track portfolios that need IRR recalculation
            
            if affected_funds:
                logger.info("🧮 Phase 3: Triggering fund-level IRR recalculations...")
                for fund_id, dates in affected_funds.items():
                    # Get portfolio_id for this fund to track portfolio-level recalculation
                    fund_info = await self.db.fetchrow(
                        "SELECT portfolio_id FROM portfolio_funds WHERE id = $1", 
                        fund_id
                    )
                    
                    if fund_info:
                        portfolio_id = fund_info["portfolio_id"]
                        affected_portfolios.add(portfolio_id)
                    
                    for date in dates:
                        await self._recalculate_fund_irr(fund_id, date)
                        result["irr_calculations"] += 1
                logger.info(f"✅ Phase 3 Complete: {result['irr_calculations']} fund IRR calculations triggered")
                
                # Phase 4: Trigger portfolio-level IRR recalculation for affected portfolios
                if affected_portfolios:
                    logger.info("🏢 Phase 4: Triggering portfolio-level IRR recalculations...")
                    for portfolio_id in affected_portfolios:
                        # Find the earliest date affected for this portfolio
                        earliest_date = await self._get_earliest_date_for_portfolio(portfolio_id, affected_funds)
                        if earliest_date:
                            portfolio_irr_count = await self._recalculate_portfolio_irr(portfolio_id, earliest_date)
                            result["portfolio_irr_recalculations"] += portfolio_irr_count
                    logger.info(f"✅ Phase 4 Complete: {result['portfolio_irr_recalculations']} portfolio IRR recalculations triggered")
            
            logger.info("🎉 Backend Transaction: All phases completed successfully")
            
        except Exception as e:
            logger.error(f"❌ Backend Transaction Error: {str(e)}")
            result["success"] = False
            result["errors"].append(str(e))
            
        return result
    
    async def _save_activity(self, activity_data: Dict[str, Any]) -> None:
        """Save a single activity to the database"""
        try:
            # Insert or update activity
            if activity_data.get('id'):
                # Update existing activity - Build dynamic update query
                set_clauses = []
                params = []
                param_count = 1
                
                for key, value in activity_data.items():
                    if key != 'id':  # Skip the id field for SET clause
                        set_clauses.append(f"{key} = ${param_count}")
                        params.append(value)
                        param_count += 1
                
                params.append(activity_data['id'])  # Add id for WHERE clause
                
                query = f"UPDATE holding_activity_log SET {', '.join(set_clauses)} WHERE id = ${param_count} RETURNING *"
                result = await self.db.fetchrow(query, *params)
            else:
                # Insert new activity - Build dynamic insert query
                columns = list(activity_data.keys())
                values = list(activity_data.values())
                placeholders = [f"${i+1}" for i in range(len(values))]
                
                query = f"INSERT INTO holding_activity_log ({', '.join(columns)}) VALUES ({', '.join(placeholders)}) RETURNING *"
                result = await self.db.fetchrow(query, *values)
            
            if not result:
                raise Exception(f"Failed to save activity: {activity_data}")
                
            logger.info(f"📥 Activity saved: {activity_data['activity_type']} for fund {activity_data['portfolio_fund_id']}")
            
        except Exception as e:
            logger.error(f"❌ Failed to save activity: {str(e)}")
            raise
    
    async def _save_valuation(self, valuation_data: Dict[str, Any]) -> None:
        """Save a single valuation to the database"""
        try:
            # Check for existing valuation
            existing_valuation = await self.db.fetchrow(
                "SELECT * FROM portfolio_fund_valuations WHERE portfolio_fund_id = $1 AND valuation_date = $2",
                valuation_data['portfolio_fund_id'],
                valuation_data['valuation_date']
            )
            
            if existing_valuation:
                # Update existing valuation - Build dynamic update query
                set_clauses = []
                params = []
                param_count = 1
                
                for key, value in valuation_data.items():
                    set_clauses.append(f"{key} = ${param_count}")
                    params.append(value)
                    param_count += 1
                
                params.append(existing_valuation['id'])  # Add id for WHERE clause
                
                query = f"UPDATE portfolio_fund_valuations SET {', '.join(set_clauses)} WHERE id = ${param_count} RETURNING *"
                result = await self.db.fetchrow(query, *params)
            else:
                # Insert new valuation - Build dynamic insert query
                columns = list(valuation_data.keys())
                values = list(valuation_data.values())
                placeholders = [f"${i+1}" for i in range(len(values))]
                
                query = f"INSERT INTO portfolio_fund_valuations ({', '.join(columns)}) VALUES ({', '.join(placeholders)}) RETURNING *"
                result = await self.db.fetchrow(query, *values)
            
            if not result:
                raise Exception(f"Failed to save valuation: {valuation_data}")
                
            logger.info(f"💰 Valuation saved: £{valuation_data['valuation']} for fund {valuation_data['portfolio_fund_id']} on {valuation_data['valuation_date']}")
            
        except Exception as e:
            logger.error(f"❌ Failed to save valuation: {str(e)}")
            raise
    
    async def _recalculate_fund_irr(self, fund_id: int, date: str) -> None:
        """Trigger IRR recalculation for a specific fund and date"""
        try:
            logger.info(f"🧮 Recalculating fund IRR for fund {fund_id} on {date}")
            
            # Use the existing IRR calculation function
            irr_result = await calculate_single_portfolio_fund_irr(
                portfolio_fund_id=fund_id,
                irr_date=date,
                db=self.db
            )
            
            if irr_result.get("success"):
                logger.info(f"✅ Fund IRR recalculated successfully for fund {fund_id}: {irr_result.get('irr_percentage', 0)}%")
            else:
                logger.warning(f"⚠️ Fund IRR recalculation failed for fund {fund_id}: {irr_result.get('error', 'Unknown error')}")
                
        except Exception as e:
            logger.error(f"❌ Fund IRR recalculation failed for fund {fund_id}: {str(e)}")
            # Don't raise - IRR calculation failure shouldn't fail the entire transaction

    async def _recalculate_portfolio_irr(self, portfolio_id: int, start_date: str) -> int:
        """
        Trigger portfolio-level IRR recalculation from a start date onwards.
        Uses the same logic as in holding_activity_logs.py
        
        Args:
            portfolio_id: The portfolio ID
            start_date: Only recalculate portfolio IRR values from this date onwards (YYYY-MM-DD format)
            
        Returns:
            The number of portfolio IRR values processed (recalculated + created)
        """
        try:
            logger.info(f"🏢 Recalculating portfolio-level IRR values for portfolio {portfolio_id} from {start_date} onwards")
            
            # Import the function from holding_activity_logs to avoid code duplication
            from app.api.routes.holding_activity_logs import recalculate_portfolio_irr_values_from_date
            
            # Call the existing portfolio IRR recalculation function
            portfolio_irr_count = await recalculate_portfolio_irr_values_from_date(
                portfolio_id, start_date, self.db
            )
            
            logger.info(f"✅ Portfolio IRR recalculation completed for portfolio {portfolio_id}: {portfolio_irr_count} values processed")
            return portfolio_irr_count
            
        except Exception as e:
            logger.error(f"❌ Portfolio IRR recalculation failed for portfolio {portfolio_id}: {str(e)}")
            # Don't raise - IRR calculation failure shouldn't fail the entire transaction
            return 0

    async def _get_earliest_date_for_portfolio(self, portfolio_id: int, affected_funds: Dict[int, List[str]]) -> Optional[str]:
        """
        Find the earliest date that affects a specific portfolio.
        
        Args:
            portfolio_id: The portfolio ID to find dates for
            affected_funds: Dictionary mapping fund_id to list of dates
            
        Returns:
            The earliest date string (YYYY-MM-DD format) or None if no dates found
        """
        earliest_date = None
        
        # Get all funds for this portfolio
        portfolio_funds = await self.db.fetch(
            "SELECT id FROM portfolio_funds WHERE portfolio_id = $1",
            portfolio_id
        )
        
        if not portfolio_funds:
            return None
        
        portfolio_fund_ids = [pf["id"] for pf in portfolio_funds]
        
        # Find the earliest date from affected funds that belong to this portfolio
        for fund_id, dates in affected_funds.items():
            if fund_id in portfolio_fund_ids:
                for date in dates:
                    if earliest_date is None or date < earliest_date:
                        earliest_date = date
        
        return earliest_date
    
    def _get_affected_funds(self, activities: List[Dict[str, Any]], valuations: List[Dict[str, Any]]) -> Dict[int, List[str]]:
        """
        Get list of funds and dates that need IRR recalculation
        
        Returns:
            Dictionary mapping fund_id to list of dates
        """
        affected_funds = {}
        
        # Process activities
        for activity in activities:
            fund_id = activity.get('portfolio_fund_id')
            date = activity.get('activity_timestamp', '').split('T')[0]
            
            if fund_id and date:
                if fund_id not in affected_funds:
                    affected_funds[fund_id] = []
                if date not in affected_funds[fund_id]:
                    affected_funds[fund_id].append(date)
        
        # Process valuations
        for valuation in valuations:
            fund_id = valuation.get('portfolio_fund_id')
            date = valuation.get('valuation_date', '').split('T')[0]
            
            if fund_id and date:
                if fund_id not in affected_funds:
                    affected_funds[fund_id] = []
                if date not in affected_funds[fund_id]:
                    affected_funds[fund_id].append(date)
        
        return affected_funds

class TransactionValidator:
    """
    Validates transaction data and ensures proper ordering
    """
    
    @staticmethod
    def validate_activities(activities: List[Dict[str, Any]]) -> List[str]:
        """Validate activity data"""
        errors = []
        
        for i, activity in enumerate(activities):
            if not activity.get('portfolio_fund_id'):
                errors.append(f"Activity {i}: Missing portfolio_fund_id")
            
            if not activity.get('activity_timestamp'):
                errors.append(f"Activity {i}: Missing activity_timestamp")
            
            if not activity.get('activity_type'):
                errors.append(f"Activity {i}: Missing activity_type")
            
            if not activity.get('amount'):
                errors.append(f"Activity {i}: Missing amount")
        
        return errors
    
    @staticmethod
    def validate_valuations(valuations: List[Dict[str, Any]]) -> List[str]:
        """Validate valuation data"""
        errors = []
        
        for i, valuation in enumerate(valuations):
            if not valuation.get('portfolio_fund_id'):
                errors.append(f"Valuation {i}: Missing portfolio_fund_id")
            
            if not valuation.get('valuation_date'):
                errors.append(f"Valuation {i}: Missing valuation_date")
            
            if valuation.get('valuation') is None:
                errors.append(f"Valuation {i}: Missing valuation amount")
        
        return errors
    
    @staticmethod
    def validate_ordering(activities: List[Dict[str, Any]], valuations: List[Dict[str, Any]]) -> List[str]:
        """Validate that activities and valuations are in proper order"""
        errors = []
        
        # Check for same fund/date combinations
        activity_fund_dates = set()
        for activity in activities:
            fund_id = activity.get('portfolio_fund_id')
            date = activity.get('activity_timestamp', '').split('T')[0]
            if fund_id and date:
                activity_fund_dates.add((fund_id, date))
        
        valuation_fund_dates = set()
        for valuation in valuations:
            fund_id = valuation.get('portfolio_fund_id')
            date = valuation.get('valuation_date', '').split('T')[0]
            if fund_id and date:
                valuation_fund_dates.add((fund_id, date))
        
        # Find overlapping fund/date combinations
        overlapping = activity_fund_dates.intersection(valuation_fund_dates)
        if overlapping:
            errors.append(f"Found {len(overlapping)} fund/date combinations with both activities and valuations - proper ordering is critical")
        
        return errors 

File: frontend/src/services/transactionCoordinator.ts
import api from './api';

export interface ActivityEdit {
  id?: number;
  fundId: number;
  month: string;
  activityType: string;
  value: string;
  isNew: boolean;
  originalActivityId?: number;
  toDelete?: boolean;
}

export interface ValuationEdit {
  id?: number;
  fundId: number;
  month: string;
  value: string;
  isNew: boolean;
  originalActivityId?: number;
  toDelete?: boolean;
}

export interface TransactionResult {
  success: boolean;
  errors: string[];
  processedActivities: number;
  processedValuations: number;
  recalculatedIRRs: number;
}

/**
 * Transaction Coordinator Service
 * Ensures activities are saved before valuations to prevent IRR calculation race conditions
 * Phase 3: Triggers IRR recalculation after all saves are complete
 */
export class TransactionCoordinator {
  
  /**
   * Save activities and valuations in correct order with IRR recalculation
   * Activities MUST be saved before valuations to ensure IRR calculations have complete data
   */
  static async saveActivitiesAndValuations(
    edits: ActivityEdit[],
    accountHoldingId: number
  ): Promise<TransactionResult> {
    
    const result: TransactionResult = {
      success: true,
      errors: [],
      processedActivities: 0,
      processedValuations: 0,
      recalculatedIRRs: 0
    };

    try {
      // Separate activities and valuations
      const activities = edits.filter(edit => edit.activityType !== 'Current Value');
      const valuations = edits.filter(edit => edit.activityType === 'Current Value');

      console.log(`🔄 Transaction Coordinator: Processing ${activities.length} activities and ${valuations.length} valuations`);

      // VALIDATION: Check activities before processing to prevent sequence waste
      if (activities.length > 0) {
        const validationErrors = this.validateBulkActivities(activities);
        if (validationErrors.length > 0) {
          throw new Error(`Validation failed: ${validationErrors.join(', ')}`);
        }
        console.log(`✅ Validation passed for ${activities.length} activities`);
      }

      // PHASE 1: Save all activities first (using bulk operations where possible)
      if (activities.length > 0) {
        console.log('📥 Phase 1: Saving activities in bulk...');
        
        await this.saveBulkActivities(activities, accountHoldingId);
        result.processedActivities = activities.length;
        
        console.log(`✅ Phase 1 Complete: ${result.processedActivities} activities saved in bulk`);
      }

      // PHASE 2: Save valuations after all activities are saved
      if (valuations.length > 0) {
        console.log('💰 Phase 2: Saving valuations...');
        
        for (const valuation of valuations) {
          await this.saveValuation(valuation);
          result.processedValuations++;
        }
        
        console.log(`✅ Phase 2 Complete: ${result.processedValuations} valuations saved`);
      }

      // PHASE 3: Trigger IRR recalculation for affected portfolio funds
      console.log('🔄 Phase 3: Triggering IRR recalculation...');
      
      // Group activities by fund and month to trigger recalculation
      const fundRecalculationMap = new Map<number, string[]>();
      
      // Process activities
      if (activities.length > 0) {
        console.log('🔄 Phase 3a: Processing activity-based IRR recalculation...');
        
        activities.forEach(activity => {
          if (!activity.toDelete) { // Only for non-deleted activities
            const fundId = activity.fundId;
            const activityDate = `${activity.month}-01`;
            
            if (!fundRecalculationMap.has(fundId)) {
              fundRecalculationMap.set(fundId, []);
            }
            fundRecalculationMap.get(fundId)!.push(activityDate);
          }
        });
      }
      
      // Process valuations - NEW: Also trigger IRR recalculation for valuation changes
      if (valuations.length > 0) {
        console.log('🔄 Phase 3b: Processing valuation-based IRR recalculation...');
        
        valuations.forEach(valuation => {
          if (!valuation.toDelete) { // Only for non-deleted valuations
            const fundId = valuation.fundId;
            const valuationDate = `${valuation.month}-01`;
            
            if (!fundRecalculationMap.has(fundId)) {
              fundRecalculationMap.set(fundId, []);
            }
            fundRecalculationMap.get(fundId)!.push(valuationDate);
            
            console.log(`📊 Valuation change detected for fund ${fundId} on ${valuationDate} - will trigger IRR recalculation`);
          }
        });
      }

      // Trigger recalculation for each affected fund (both activities and valuations)
      if (fundRecalculationMap.size > 0) {
        for (const [fundId, dates] of fundRecalculationMap) {
          try {
            // Get the earliest date for this fund to trigger recalculation
            const earliestDate = dates.sort()[0];
            
            console.log(`🔄 Triggering IRR recalculation for fund ${fundId} from date ${earliestDate}`);
            
            const recalcResponse = await api.post(`portfolio_funds/${fundId}/recalculate-irr`, {
              activity_date: earliestDate
            });
            
            if (recalcResponse.data?.recalculation_result?.recalculated_existing) {
              const recalculatedCount = recalcResponse.data.recalculation_result.recalculated_existing;
              result.recalculatedIRRs += recalculatedCount;
              console.log(`✅ Recalculated ${recalculatedCount} IRR values for fund ${fundId}`);
            } else {
              console.log(`✅ IRR recalculation triggered for fund ${fundId} (no existing IRRs to recalculate)`);
            }
            
          } catch (error: any) {
            console.warn(`⚠️ IRR recalculation failed for fund ${fundId}:`, error.message);
            // Don't fail the entire transaction for IRR recalculation issues
          }
        }
        
        console.log(`✅ Phase 3 Complete: ${result.recalculatedIRRs} IRR values recalculated across ${fundRecalculationMap.size} funds`);
      } else {
        console.log('✅ Phase 3 Complete: No IRR recalculation needed (no non-deleted activities or valuations)');
      }

      console.log('🎉 Transaction Coordinator: All saves completed successfully');
      
    } catch (error: any) {
      console.error('❌ Transaction Coordinator Error:', error);
      result.success = false;
      result.errors.push(error.message || 'Unknown error occurred');
    }

    return result;
  }

  /**
   * Save a single activity
   */
  private static async saveActivity(activity: ActivityEdit, accountHoldingId: number): Promise<void> {
    const backendActivityType = this.convertActivityTypeForBackend(activity.activityType);
    
    const activityData = {
      portfolio_fund_id: activity.fundId,
      product_id: accountHoldingId,
      activity_type: backendActivityType,
      activity_timestamp: `${activity.month}-01`,
      amount: activity.value
    };

    console.log(`📥 Saving activity: ${JSON.stringify(activityData)}`);

    if (activity.toDelete && activity.originalActivityId) {
      await api.delete(`holding_activity_logs/${activity.originalActivityId}`);
    } else if (activity.isNew) {
      // Skip IRR calculation for new activities - it will be calculated after valuation
      await api.post('holding_activity_logs', activityData, { 
        params: { skip_irr_calculation: true } 
      });
    } else if (activity.originalActivityId) {
      await api.patch(`holding_activity_logs/${activity.originalActivityId}`, activityData);
    }
  }

  /**
   * Save activities in bulk using sequence reservation
   * This replaces individual API calls with a single bulk operation for new activities
   */
  private static async saveBulkActivities(
    activities: ActivityEdit[], 
    accountHoldingId: number
  ): Promise<void> {
    
    if (activities.length === 0) return;
    
    console.log(`🚀 BULK: Saving ${activities.length} activities with sequence reservation`);
    
    // Separate activities by operation type
    const newActivities = activities.filter(activity => activity.isNew && !activity.toDelete);
    const updateActivities = activities.filter(activity => !activity.isNew && !activity.toDelete && activity.originalActivityId);
    const deleteActivities = activities.filter(activity => activity.toDelete && activity.originalActivityId);
    
    // Handle bulk creation for new activities
    if (newActivities.length > 0) {
      console.log(`📥 BULK: Creating ${newActivities.length} new activities...`);
      
      // Convert to backend format
      const activityData = newActivities.map(activity => ({
        portfolio_fund_id: activity.fundId,
        product_id: accountHoldingId,
        activity_type: this.convertActivityTypeForBackend(activity.activityType),
        activity_timestamp: `${activity.month}-01`,
        amount: parseFloat(activity.value)
      }));

      // Use bulk endpoint with sequence reservation
      const response = await api.post('holding_activity_logs/bulk', activityData, {
        params: { skip_irr_calculation: true }
      });
      
      console.log(`✅ BULK: Successfully created ${response.data.length} activities`);
    }
    
    // Handle individual updates (can't be bulked due to different IDs)
    if (updateActivities.length > 0) {
      console.log(`🔄 BULK: Updating ${updateActivities.length} existing activities individually...`);
      for (const activity of updateActivities) {
        await this.saveActivity(activity, accountHoldingId);
      }
    }
    
    // Handle individual deletions (can't be bulked due to different IDs)
    if (deleteActivities.length > 0) {
      console.log(`🗑️ BULK: Deleting ${deleteActivities.length} activities individually...`);
      for (const activity of deleteActivities) {
        await this.saveActivity(activity, accountHoldingId);
      }
    }
  }

  /**
   * Validate activities before bulk save to prevent sequence reservation waste
   */
  private static validateBulkActivities(activities: ActivityEdit[]): string[] {
    const errors: string[] = [];
    
    activities.forEach((activity, index) => {
      const activityNum = index + 1;
      
      if (!activity.fundId) {
        errors.push(`Activity ${activityNum}: Missing fund ID`);
      }
      
      if (!activity.activityType) {
        errors.push(`Activity ${activityNum}: Missing activity type`);
      }
      
      if (!activity.value || isNaN(parseFloat(activity.value))) {
        errors.push(`Activity ${activityNum}: Invalid amount value: ${activity.value}`);
      }
      
      if (!activity.month || !/^\d{4}-\d{2}$/.test(activity.month)) {
        errors.push(`Activity ${activityNum}: Invalid month format (expected YYYY-MM): ${activity.month}`);
      }
    });
    
    return errors;
  }

  /**
   * Save a single valuation
   */
  private static async saveValuation(valuation: ValuationEdit): Promise<void> {
    const valuationData = {
      portfolio_fund_id: valuation.fundId,
      valuation_date: `${valuation.month}-01`,
      valuation: parseFloat(valuation.value)
    };

    console.log(`💰 Saving valuation: ${JSON.stringify(valuationData)}`);

    if (valuation.toDelete && valuation.originalActivityId) {
      await api.delete(`fund_valuations/${valuation.originalActivityId}`);
    } else if (valuation.isNew) {
      await api.post('fund_valuations', valuationData);
    } else if (valuation.originalActivityId) {
      await api.patch(`fund_valuations/${valuation.originalActivityId}`, valuationData);
    }
  }

  /**
   * Convert UI activity type to backend format
   */
  private static convertActivityTypeForBackend(activityType: string): string {
    const typeMap: { [key: string]: string } = {
      'Investment': 'Investment',
      'Withdrawal': 'Withdrawal',
      'Switch In': 'Switch In',
      'Switch Out': 'Switch Out',
      'Fee': 'Fee',
      'Interest': 'Interest',
      'Dividend': 'Dividend',
      'Capital Gain': 'Capital Gain',
      'Capital Loss': 'Capital Loss',
      'Other': 'Other'
    };
    
    return typeMap[activityType] || activityType;
  }

  /**
   * Group edits by month and fund for batch processing
   */
  static groupEditsByMonthAndFund(edits: ActivityEdit[]): Map<string, ActivityEdit[]> {
    const groups = new Map<string, ActivityEdit[]>();
    
    for (const edit of edits) {
      const key = `${edit.month}_${edit.fundId}`;
      if (!groups.has(key)) {
        groups.set(key, []);
      }
      groups.get(key)!.push(edit);
    }
    
    return groups;
  }
} 